velox/functions/prestosql/Bitwise.h

```diff
@@ -95,67 +95,79 @@ struct BitwiseArithmeticShiftRightFunction {
+  if ((uint32_t)shift >= kMaxShift) {
+    return 0;
   } else {
-    result = (number << shift);
+    return (number << shift);
   }
-  return true;
 }
-} // namespace
+} // namespace detail
 
-template <typename T>
+template <typename TExec>
 struct BitwiseLeftShiftFunction {
-  template <typename TInput>
-  FOLLY_ALWAYS_INLINE bool call(int64_t& result, TInput number, TInput shift) {
-    if constexpr (std::is_same_v<TInput, int8_t>) {
-      return bitwiseLeftShift<TInput, 8>(result, number, shift);
-    } else if constexpr (std::is_same_v<TInput, int16_t>) {
-      return bitwiseLeftShift<TInput, 16>(result, number, shift);
-    } else if constexpr (std::is_same_v<TInput, int32_t>) {
-      return bitwiseLeftShift<TInput, 32>(result, number, shift);
-    } else {
-      return bitwiseLeftShift<TInput, 64>(result, number, shift);
-    }
+  FOLLY_ALWAYS_INLINE void
+  call(int64_t& result, int64_t number, int32_t shift) {
+    result = detail::bitwiseLeftShift<int64_t>(number, shift);
+  }
+
+  FOLLY_ALWAYS_INLINE void
+  call(int32_t& result, int32_t number, int32_t shift) {
+    result = detail::bitwiseLeftShift<int32_t>(number, shift);
+  }
+
+  FOLLY_ALWAYS_INLINE void
+  call(int16_t& result, int16_t number, int32_t shift) {
+    result = detail::bitwiseLeftShift<int16_t>(number, shift);
+  }
+
+  FOLLY_ALWAYS_INLINE void call(int8_t& result, int8_t number, int32_t shift) {
+    result = detail::bitwiseLeftShift<int8_t>(number, shift);
   }
 };
 
-namespace {
-template <typename T, typename UT, int MAX_SHIFT>
-FOLLY_ALWAYS_INLINE bool bitwiseRightShift(int64_t& result, T number, T shift) {
-  if ((uint64_t)shift >= MAX_SHIFT) {
-    result = 0;
+namespace detail {
+template <typename T, typename UT>
+FOLLY_ALWAYS_INLINE T bitwiseRightShift(T number, int32_t shift) {
+  static constexpr uint32_t kMaxShift = sizeof(T) * 8;
+  if ((uint32_t)shift >= kMaxShift) {
+    return 0;
   } else {
-    result = ((UT)number >> shift);
+    return ((UT)number) >> shift;
   }
-  return true;
 }
-} // namespace
+} // namespace detail
 
-template <typename T>
+template <typename TExec>
 struct BitwiseRightShiftFunction {
-  template <typename TInput>
-  FOLLY_ALWAYS_INLINE bool call(int64_t& result, TInput number, TInput shift) {
-    if constexpr (std::is_same_v<TInput, int8_t>) {
-      return bitwiseRightShift<int8_t, uint8_t, 8>(result, number, shift);
-    } else if constexpr (std::is_same_v<TInput, int16_t>) {
-      return bitwiseRightShift<int16_t, uint16_t, 16>(result, number, shift);
-    } else if constexpr (std::is_same_v<TInput, int32_t>) {
-      return bitwiseRightShift<int32_t, uint32_t, 32>(result, number, shift);
-    } else {
-      return bitwiseRightShift<int64_t, uint64_t, 64>(result, number, shift);
-    }
+  FOLLY_ALWAYS_INLINE void
+  call(int64_t& result, int64_t number, int32_t shift) {
+    result = detail::bitwiseRightShift<int64_t, uint64_t>(number, shift);
+  }
+
+  FOLLY_ALWAYS_INLINE void
```

Comment:
same comment, maybe just provide one templated version of it, then let it get instantiated based on the registered types

Reply:
I'm not sure I know how to generate 'unsigned' version of T. Do you know?

Reply:
Looks like I can use https://en.cppreference.com/w/cpp/types/make_unsigned

Let me try.

Reply:
It seems to work and the code is much shorter. Thank you for the suggestion.

```
template <typename TExec>
struct BitwiseRightShiftFunction {
  template <typename T>
  FOLLY_ALWAYS_INLINE void call(T& result, T number, int32_t shift) {
    static constexpr uint32_t kMaxShift = sizeof(T) * 8;

    // Return zero if 'shift' is negative or exceeds number of bits in T.
    if ((uint32_t)shift >= kMaxShift) {
      result = 0;
    } else {
      result = ((std::make_unsigned_t<T>)number) >> shift;
    }
  }
};
```